# Part 1 - Python basics

In Python, everything is an object. You can think of an object as a box into which you can put different things. Another name for a box is a **variable**. Variables allow us to store various data types and keep out code more organized.

#### Variables

In [1]:
# To assign a variable we use an assignment operator =
x = 5 # Assigned 5 to x. Now x contains 5 in it.

In [2]:
# Let's confirm it by printing x
print(x)

5


In [3]:
# It can also contain single characters
x = 'a'
print(x)

a


In [4]:
# Sentences e.g. strings
x = "Hello, this is a sentence!"
print(x)

Hello, this is a sentence!


In [5]:
# Arithmetic formulas
x = 2 + 5
print(x)

7


Virtually anything. Notice that every time we assigned a new value to the variable x, the old value got overwritten.

#### Numbers

Python allows us to perform basic arithmetic just on its own.

In [6]:
1 + 1

2

In [7]:
2 - 1

1

In [8]:
3 * 4

12

In [9]:
15 / 3

5.0

In [10]:
3 ** 2 # 3 in the power of 2

9

In [11]:
5 % 4 # % is called modulo. It is the remainder that is left after division. If we divide 5 by 4 we get 1 and 1/4.
# In 1/4, 1 is the result of modulo

1

In [12]:
10 % 8

2

In [13]:
3 % 2 

1

In [14]:
10 % 5

0

In [15]:
(2 + 3) * (5 + 5)

50

Generally, there are two types of numbers. Whole numbers called integers (int), numbers with decimal points called floats (float).

In [16]:
type(1)

int

In [17]:
type(1.0)

float

In [18]:
# Notice that 15/3 in the example above gives us a float
x = 15/3
type(x)

float

It is important to keep in mind what kind of numbers you are working with, as in some cases you might be working with integers, but expecting a float and vice versa. For example:

In [19]:
15/4

3.75

Gives us 3.75. However, you might not always get the same answer depending on Python version and IDE you are using. In some cases it is possible to get 3 as an answer. This might happen because of "classic" division. In classic division, if you divide an integer by an integer, the answer you get is also an integer.

In [20]:
int(15/4)

3

Python 3 on the other hand performs "true" division, giving you the result you would expect, even if you do not explicitly declare it.

In [21]:
15/4

3.75

In [22]:
# It is generally a good practice to explicitly state that a number is a float if you expect a float as an answer.
15.0/4.0
# By adding .0 to the end, we convert an integer to a float

3.75

In [23]:
# Furthermore, converting only one number to a float will automatically convert the answer to a float as well.
15.0/4

3.75

In [24]:
# Same as
15/4.0

3.75

In [25]:
# Same as 
15.0/4.0
# I'd recommend adding .0 to all numbers to keep things consistent

3.75

#### Strings

String is a collection of individual characters. Anything inside of 'single quotes' or "double quotes" is a string.

In [26]:
type('a') # Single character is also a string in Python

str

In [27]:
'single quotes'

'single quotes'

In [28]:
"double quotes"

'double quotes'

Double quotes are preferred when there a single quote exists in a sentence. This allows Python avoid confusion.

In [29]:
"It's a beautiful day"

"It's a beautiful day"

In [30]:
# If you try to use single quotes in the example above, Python will think that the string ends before s
'It's a beautiful day 

SyntaxError: invalid syntax (<ipython-input-30-868252902d2f>, line 2)

In [31]:
# Not the result we are expecting

#### Lists
List is a collection of variables. It allows us to store multiple variables in a single variable. So instead of assigning multiple values to multiple variables, we can assign multiple values to a single variable. For example:

In [32]:
# Assigning:
x1 = 1
x2 = 2
x3 = 3
# is burdensome. It get's pretty messy if we need to initiate a large number of variables, let's say a 100

In [33]:
# Much better to use list. Square brackets let Python know that the object is a list:
[1,2,3]

[1, 2, 3]

In [34]:
# Like with any other object, we can assign list to a variable
my_list = ['A', 2, 'Hello'] # It is possible to have multiple data types (like strings and numbers) in the same list
print(my_list)

['A', 2, 'Hello']


In [35]:
# It is also possible to have a list within a list
my_list = ['A', 2, 'Hello',['Another','list']] # Because my_list has another list in it, it's called a nested list
print(my_list)

['A', 2, 'Hello', ['Another', 'list']]


#### List indexing

You can access individual elements by using [] next to the name. In Python, an index starts with 0, so if we want to access the first element of a list, we need to pass 0. General formula is if we want to access Nth element of a list, we need to ask for position located at N - 1. Imagine a list with 200 different values in it. If we want to access value 100, we need to pass [99], value 168 is located in the position [167].

In [36]:
my_list[0] # First element

'A'

In [37]:
my_list[2] # Third element

'Hello'

In [38]:
big_list = [x for x in range(1,201)] # Don't worry if you don't undersant this expression. We will cover list comprehensions
# later in this course

In [39]:
print(big_list)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200]


In [40]:
big_list[99]

100

In [41]:
big_list[167]

168

In [42]:
# Recall that my_list has another list within it
print(my_list)

['A', 2, 'Hello', ['Another', 'list']]


In [43]:
my_list[3]

['Another', 'list']

In [44]:
# But how can we access individual values in that second list? We can do that by using second pair of square brackets []
my_list[3][0]

'Another'

In [45]:
my_list[3][1]

'list'

So in this case, we have two strings in the inner list. We can actually access individual characters in a string by doing the same operation.

In [46]:
"This is a string"[0] # Prints T because we are accessing the first element of that string e.g. the first letter in the sentence

'T'

In [47]:
"This is a string"[4] # Just a space

' '

In [48]:
"This is a string"[5]

'i'

In [49]:
# We can do the same thing in our list
my_list[3][0]

'Another'

In [50]:
my_list[3][0][0]

'A'

In [51]:
my_list[3][0][1]

'n'

In [52]:
my_list[3][1]

'list'

In [53]:
my_list[3][1][3]

't'

#### Slicing
Slicing means you extract a certain part of a list, string, data frame, etc... Some objects can be slices, while others can not. 

In [54]:
# Using our exmaple above
"This is a string"[:4] # Grab all characters up to, but not including, character number 5 (remember index starts with 0)

'This'

In [59]:
"This is a string"[:12]

'This is a st'

In [60]:
"This is a string"[3:9]

's is a'

In [61]:
"This is a string"[5:] # Grab all characters, starting from position 5

'is a string'

In [62]:
"This is a string"[::2] # Grab every second character

'Ti sasrn'

In [63]:
"This is a string"[::-1] # Print backwards

'gnirts a si sihT'

In [64]:
"This is a string"[::-2] # Print backwards every second character

'git  ish'

In [65]:
# The general syntax is [start:end:step]
"This is a string"[2:12:3] # Grab characters, starting from index 2 to 12, in the step of 3

'iiat'

In [66]:
# The same can be done if we assign string to a variable
s = "This is a string"
s[3:8]

's is '

In [69]:
# Or we can assign sliced version of a string to a variable
s1 = "This is a string"[1:4]
s1

'his'

In [73]:
# Same can be done with lists
my_list

['A', 2, 'Hello', ['Another', 'list']]

In [74]:
my_list[2:]

['Hello', ['Another', 'list']]

In [88]:
my_list[3] # Grabbing the nested list ['Another', 'list']

['Another', 'list']

In [89]:
my_list[3][0] # Grabbing 'Another'

'Another'

In [90]:
my_list[3][0][:4] # Grabbing the first 4 characters

'Anot'

#### Other data types
**Tuples** are lists that can not be modified. We can initiate a tuple by using round brackets ().

In [91]:
my_tuple = () # Create an empty tuple

In [94]:
my_tuple = (1,2,3,4,5,'hi') # Can contain various data types like lists
my_tuple

(1, 2, 3, 4, 5, 'hi')

In [96]:
# But can not be modified
my_tuple[2] = 4

TypeError: 'tuple' object does not support item assignment

In [98]:
# Unlike lists
my_list[1]

2

In [102]:
my_list[1] = 2345
my_list[1]

2345

In [103]:
my_list

['A', 2345, 'Hello', ['Another', 'list']]

**Dictionary** is exactly what it sounds like. Python dictionaries are similar to paper dictionaries. When you open a dictionary, you will find words and their definitions. In Python, words are called keys and definitions are called values. Let's look at an example of a dictionary:

In [104]:
my_dict = {} # Create dictionary using curly brackets

In [106]:
type(my_dict)

dict

In [108]:
# We can create a dictionary with initial values as:
my_dict = {'key':'value'}
my_dict

{'key': 'value'}

In [109]:
# Now we can call key to get the value
my_dict['key']

'value'

In [110]:
my_dict[0] # Using traditional indexing will result in an error, as there's no such key '0' in my_dict

KeyError: 0

In [112]:
# The way we can add new keys to my_dict is very easy:
my_dict['New Key'] = 'New Value'
my_dict

{'New Key': 'New Value', 'key': 'value'}

In [113]:
my_dict['Name'] = 'Ed'
my_dict['Age'] = 32 

In [114]:
my_dict

{'Age': 32, 'Name': 'Ed', 'New Key': 'New Value', 'key': 'value'}

In [116]:
# The length of a dictionary is defined by the number of keys it has. In our case we have 4 keys, therefore the length is 4.
len(my_dict)

4

Both tuples and dictionaries can be extremely useful in certain situations. Without going much into details, tuples are useful when you do not want any program/user to change the values of a list, so you create a tuple instead. Dictionaries are useful in a number of ways. One way dictionary is useful is if you want to look at two lists at the same time. Let's say you have list_1 containing website URLs and second list containing their description. It is very convenient to create a dictionary to store the keys and values, instead of looking into two different lists.

### Functions
You can create your own functions by using def keyword. There are a couple of things that you need to know about functions. First is that functions can, but don't have to, take arguments. Whether you need to pass an argument or not depends on your function (more about it later). Second, functions can return values. To return a value use return keyword. Finally, variables in functions exist only with a function, unless you return them. They are called local variables (as opposed to global variable). It is easier to understand functions by using examples.

In [117]:
# Let's create a function that prints 'Hello World!'
def hello_world():
    print('Hello World!')

In [118]:
hello_world()

Hello World!


In [119]:
hello_world

<function __main__.hello_world>

So what did just happen? First of all, we created a function that we decided to name hello_world. Also, we did not pass any arguments in the round brackets next to the name, as we didn't need to. Third, the function simply printed out 'Hello World!' without returning any values. Finally, after creating the function, we had to call it by using its name and round brackets(). Note that if you simply run hello_world, Python will not execute the function, but instead simply point to the object. Let's look at another example. In this case, we will create a function that sums two numbers and returns the result.

In [129]:
def sum_two_nums(a, b):
    return a + b

In [130]:
number1 = 4
number2 = 5
sum_two_nums(number1, number2)

9

Okay, let's break down what we just did. Firstly, we created a function that sums two numbers. Because it sums two numbers, we need to pass two arguments to the Python, to let it know what numbers to sum. The function returns the result of a + b. Notice that different names were used in the declaration of the function sum_two_nums(a, b) and during the execution of the function sum_two_nums(number1, number2). Why Python doesn't get confused and raises an error? That's because a and b are local variables, meaning that only that function can see and use them. If we analyze the process step by step, what we will see is that:<br>
1) 4 is assigned to number1 and 5 is assigned to number2, then<br>
2) The function sum_two_nums is called and two arguments (number1, number2) are being passed. At this moment, number1 becomes a and number2 becomes b.<br>
3) The function starts working with a and b<br>
4) It returns the result of a + b

In [124]:
# If we try to call a, then we will get an error as a only exists inside of the function, and not anywhere else
a

NameError: name 'a' is not defined

In [125]:
# Same for b
b

NameError: name 'b' is not defined

In [134]:
# Let's look at another example. In this case, we will create a function that converts celsius to fahrenheit by taking user input
def cel_to_fahr():
    user_input = input("Enter the temperature: ")    
    # Formula to convert is: (°C × 9/5) + 32 = °F
    result = (user_input * 9/5) + 32
    print(result)

In [135]:
# Let's try to run it
cel_to_fahr()

Enter the temperature: 24


TypeError: unsupported operand type(s) for /: 'str' and 'int'

Oops, why do we get an error? This is because the result of input function is a string. Even though we have entered a number, the number was converted to a string. We haven't covered conversion yet, but it is a pretty simple thing to do. Data types can be converted into other data types as long as it makes sense to do. In this case, input gives us a string, so we want to convert it to an integer. We can do it by using int() function.

In [138]:
'35'

'35'

In [136]:
type('35') # <- is a string

str

In [139]:
int('35') # <- now it's a number

35

In [140]:
type(int('35'))

int

Therefore, input() function needs to be converted to an int. Otherwise, we can't use it.

In [151]:
def cel_to_fahr():
    user_input = int(input("Enter the temperature: "))
    # Formula to convert is: (°C × 9/5) + 32 = °F
    result = (user_input * 9/5) + 32
    print(result)

In [142]:
cel_to_fahr()

Enter the temperature: 24
75.2


Let's break down what happened. Even though we didn't pass any arguments, the function uses user input to generate its result. After calling the function, the user needs to input a number, after which the number is converted to an integer and is used to calculate the result. Finally, the function prints the result. We can modify it to take user input as an argument by doing the following:

In [145]:
def cel_to_fahr_2(inp):
    '''
    This function converts celsius to fahrenheit.
    '''
    # Formula to convert is: (°C × 9/5) + 32 = °F
    result = (inp * 9/5) + 32
    print(result)

In [146]:
user_input_2 = int(input("Enter the temperature: "))
cel_to_fahr_2(user_input_2)

Enter the temperature: 24
75.2


Either works. Notice that I added a docstring to explain what the function does. 

In [152]:
help(cel_to_fahr)

Help on function cel_to_fahr in module __main__:

cel_to_fahr()



In [153]:
help(cel_to_fahr_2)

Help on function cel_to_fahr_2 in module __main__:

cel_to_fahr_2(inp)
    This function converts celsius to fahrenheit.



In [154]:
# As a little excercise try to write a function that takes hours provided by a user and converts it to seconds

## The End